### ISSUE TITLE
Proposal: only use UTC time for all back-end processing, back-end cli requests

### ISSUE THREAD

**Issue by jzohrab** [covidatlas/li#80]
Currently, the system uses timezone-aware code for all back end and CLI requests.  This has resulted in a few odd behaviours:

* CI tests crawling data at one date, and then missing cache immediately after when trying to scrape
* different data being scraped when a date is specified on the CLI

(These may have been fixed!)

This could lead to potential ambiguity when dealing with code.  For example, if a source's scraper is specified as having a startDate of '2020-05-02', that could change depending on if it's the locale date, or UTC.

This also complicates code; e.g. in load-cache/index.js, we have to cast dates to locale-specific dates.

I feel that all of these complications could be avoided if we simply specify that all back end processing is **UTC-only**:

* Calling a scrape for `2020-05-02` should only hit cache files in the folder `[location-id]/2020-05-02`
* Calling a crawl on a date should create cache files in the correct UTC folder
* Specifying a startDate of `2020-05-03` for a scraper should be interpreted as `2020-05-03T00:00:00.000Z`

This is for back-end processing only.  Human users should be able to put in things that make sense for them -- e.g. a Singapore user may be at a different date than one from San Diego -- but that is display only, and not related to back-end processing.

* How will this make life better?

No confusion, more consistency

* How big is this change (small/med/large/really large)? [Size]

Probably not much for current code, we'd just remove any locale-datetime adjusting.

* How complicated is it? / What areas does it touch?

Not sure

* What's the test plan?

points above: crawl and scrape should be consistent and respect the times. :-)

We _could_ create a CLI flag like `--utc-only`, but that feels like an unnecessary patch and still leaves the confusing code paths present in the back end.

**Comment by ryanblock** [comment]
<img src="https://user-images.githubusercontent.com/200964/80934479-41a54500-8d7d-11ea-9f31-051af27297e4.png" width=400>

> CI tests crawling data at one date, and then missing cache immediately after when trying to scrape

This was resolved in https://github.com/covidatlas/li/commit/a9de7acae5bc46251eed0c55cfe6f7f604375f29

> different data being scraped when a date is specified on the CLI

This is a theoretical issue; it only impacts source maintainers at a moment in time where the source locale they're working on has crossed the dateline, and they have not. To the best of my knowledge this has not been an issue in practice, and our logging makes it pretty clear when this comes up, but please let me know if I've got the wrong of that.

> This could lead to potential ambiguity when dealing with code. For example, if a source's scraper is specified as having a startDate of '2020-05-02', that could change depending on if it's the locale date, or UTC.

Seems like a pretty easy thing to disambiguate. (For the record, `startDate` assumes the date of the locale in question, so the developer does not have to worry about doing UTC offset calculations when setting a `startDate`.)

> This also complicates code; e.g. in load-cache/index.js, we have to cast dates to locale-specific dates.

Again, if the assertion is that building global products that are aware of timezone are more complicated than building products that ONLY work in UTC and have absolutely no idea about the concepts of timezones, then yes, I totally agree. But that's not this project – we don't get to opt out of that, unfortunately.

However, we DO get to decide where and how we manage this necessary complexity. Our current implementation completely abstracts all timezone code away from developing sources, frontend, our soon-to-ship API, and almost all regular backend work. I'd say that's a pretty tidy approach! But if anyone would like to demonstrate a simpler approach that doesn't put the onus on sources / frontend to solve the problems we've already solved, walking through point by point, end to end, how to resolve data to the dates of our official sources, I'm here for it!

> I feel that all of these complications could be avoided if we simply specify that all back end processing is UTC-only:

There may be a misconception; backend processing is, in fact, UTC-based. Our caches are UTC based. We normalize to UTC, as we must. Only when we need to engage in persistence operations do we locale-cast it, and this is by necessity: so our data resolves exactly to the dates of the official sources. Do you mean say that you'd like to specify that backend developer interfaces for dates are UTC-based?

> Calling a scrape for 2020-05-02 should only hit cache files in the folder [location-id]/2020-05-02

Why? How does this benefit us?

> Calling a crawl on a date should create cache files in the correct UTC folder

Crawler cannot and is never called on a date; every new crawl saves UTC cache files to the correct UTC folder – deep under the hood everything is UTC normalized, this is already how the system works.

> Specifying a startDate of 2020-05-03 for a scraper should be interpreted as 2020-05-03T00:00:00.000Z

I think this sounds like a potential developer interface concern.

> This is for back-end processing only. Human users should be able to put in things that make sense for them -- e.g. a Singapore user may be at a different date than one from San Diego -- but that is display only, and not related to back-end processing.

I'm not following – that's already how it works.

> How will this make life better?
> No confusion, more consistency

I'd love to continue to help dispel any confusion about this; things are completely consistent in implementation, however. If there is anything inconsistent, please do show me where so that it can be resolved.

> How big is this change (small/med/large/really large)? [Size]
> Probably not much for current code, we'd just remove any locale-datetime adjusting.

I don't know specifically what is being proposed here (see previous comments), but anything related to how the system understands and acts on dates is not a small or trivial change. The current system is reduced to essential and necessary complexity only and meets our requirements for data accuracy and consistency, while ensuring things are ergonomic for developing against our sources, frontend, and soon our API.

> What's the test plan?
> points above: crawl and scrape should be consistent and respect the times. :-)

The crawler only works in UTC. It crawls, and saves to UTC; the scraper reads UTC files and picks the freshest one for its locale – that's it!

In conclusion:
<img src="https://user-images.githubusercontent.com/200964/80934767-854c7e80-8d7e-11ea-84e6-b88685ed37c7.png" width=400>

**Comment by jzohrab** [comment]
Having read all the above, a bunch of answers:

> Do you mean say that you'd like to specify that backend developer interfaces for dates are UTC-based?"

Yes, that's the title of the issue :-) : "only use UTC time for all back-end processing, back-end cli requests".  Locale-specific info only belongs at the front end.

> while ensuring things are ergonomic for developing against our sources, frontend, and soon our API.

That could be the crux.  Ergonomic vs ... datanomic?  Systonomic?

> For the record, startDate assumes the date of the locale in question.

That certainly wasn't clear to me!  I feel it's mixing back-end processing (which should use UTC) from front-end displaying.

> But that's not this project – we don't get to opt out of that, unfortunately.

I think that do get to opt out of it, for the back end.  Not the front end, of course.

> "Calling a scrape for 2020-05-02 should only hit cache files in the folder [location-id]/2020-05-02"  -- Why? How does this benefit us?

Simplicity and obviousness.  If I'm a dev calling a command line with a param, or a job calling with a param, I would expect that param to be used, and not modified.  An example where back-end processing (CI) is failing b/c of timezone problems: I have cache for a source:

```
$ ls crawler-cache/au-au-from-wa-health/
2020-05-02
```

and the test calls scrape ("Scrape: au-au-from-wa-health / 2020-05-02"), but it fails with

```
    ✖ au-au-from-wa-health cache scrape 2020-05-02 failed: Error: Could not load cache; cache not available locally or in S3 (fail at: tests/integration/shared/sources/new-or-changed-sources-test.js:169:9)
```

For another example of added complexity, events/scraper/load-cache/_get-dated-folders.js gets folders from yesterday, today, and today.  If we assumed that all back-end requests are UTC only, there's no need to add that.

> But if anyone would like to demonstrate a simpler approach that doesn't put the onus on sources / frontend to solve the problems we've already solved, walking through point by point, end to end, how to resolve data to the dates of our official sources, I'm here for it!

There is something I must be missing in the question.  I would just submit a PR removing any Tz considerations from the code, for crawling and scraping:

* For non-timeseries, since our crawl and scrapes are always point-in-time events, storing them as UTC means that we don't need to worry about date changes.
* For timeseries, if the source pages actually reference _dates and times_ in their data, then things do get tricky ... e.g., a URL might be crawled at 05-04 UTC, but could contain data saying "on 05-05 [local date/time] cases = 42".  That would be up to the scraper to deal with, though.

Front-end will always have an onus on dealing with the Tz issues, this question is only about back-end.

I'm not clear what I'm missing for "resolve data to the dates of our official sources."  We're storing data in UTC, so resolving to their date means that the API accepts their Tz as a param, and they get the right data back ... or they have a separate API.

Thanks for the long response, cheers!

**Comment by ryanblock** [comment]
Above I said this:

> ... if the assertion is that building global products that are aware of timezone are more complicated than building products that ONLY work in UTC and have absolutely no idea about the concepts of timezones, then yes, I totally agree. But that's not this project – we don't get to opt out of that, unfortunately.
> However, we DO get to decide where and how we manage this necessary complexity.

The complexity of working with timezones lives in our system; what I hear you proposing is shifting that complexity from the area it it lives in today to other, different areas. This is something we can do.

However, this will result in a large overall increase in complexity, developer cognitive overhead, and bug surface area, especially as it pertains to building anything that relates to developing web views, API endpoints, and sources.

For example: we currently locale-cast (my term for UTC denormalization) our data as a scrape occurs and is persisted into the database. What you're proposing is that shift that process to denormalizing UTC on the way out of the database, so in our API logic, and with client side browser datetime libraries, etc.

Would it not be better to run this code in one place at one time in `scraper`, instead of running it in, say, all of our various API endpoints that would necessarily need to run this process millions of times servicing all our requests?

So again, the complexity remains. We're just talking about where we want it to go. I prefer to have it one place, surrounded by tests, deterministic, and out of everyone else's way. If I were developing and maintaining sources, I would not want this to be amongst my considerations.

Consider this goal state:
- On April 1st, the city and county of San Francisco says there are 1000 cases
- `covidatlas.com/san-francisco` shows 1000 cases for April 1st
- `api.covidatlas.com/san-francisco` shows 1000 cases for April 1st
- The source dev that runs `./start --crawl us-ca-san-francisco-county --date 2020-04-01` gets a result of `1000` cases

> For non-timeseries, since our crawl and scrapes are always point-in-time events, storing them as UTC means that we don't need to worry about date changes.

I assume by `date changes` you're referring to any moment where where the dateline has been crossed by UTC compared to the current local date/time; if so, that statement is incorrect. In order to meet the above standard for our data, we have to know where the location is in order to denormalize UTC. We don't get to opt out of the complexity, we just get to decide where it goes. One place, in one way, or many places in many ways?

> For timeseries, if the source pages actually reference dates and times in their data, then things do get tricky ... e.g., a URL might be crawled at 05-04 UTC, but could contain data saying "on 05-05 [local date/time] cases = 42". That would be up to the scraper to deal with, though.

Strongly disagree; scrapers are not in charge of data persistence and this is what I mean about shifting the complexity from where it lives today onto other developers within our ecosystem. I don't see why we should be relying on / asking source developers to deal with the intersection of the trickiest computer problems: timezones and data persistence. That absolutely should not be a scraper consideration.

> Front-end will always have an onus on dealing with the Tz issues, this question is only about back-end.

This is untrue; with our current system design, frontend and API do not have any issue achieving the goal state outlined above, and can do so without any UTC denormalization, (client-side) libraries, etc. It just works.

Before we go further on this issue, let me ask: Is there something that currently does not work that should be working? Is there something broken that needs fixing? Or are we vetting general dissatisfaction with the specific location that timezone complexity currently resides within our system?

**Comment by jzohrab** [comment]
We have a new flag in master `--utc`, which seems to work and addresses issues I was having in CI.  This appears to work fine, and avoids date/time adjustments.  🎉  👍  Thanks!
